test/WeakPartialOrderingTest.cpp

```diff
@@ -0,0 +1,853 @@
+      EXPECT_EQ(answer.exit, wpo.is_exit(v)) << wpo.get_node(v);
+      EXPECT_EQ(answer.num_succs, wpo.get_successors(v).size())
+          << wpo.get_node(v);
+      EXPECT_EQ(answer.num_preds, wpo.get_num_preds(v)) << wpo.get_node(v);
+      EXPECT_EQ(answer.widen, wpo.is_widening_point(v)) << wpo.get_node(v);
+      if (wpo.is_head(v)) {
+        EXPECT_EQ(wpo.get_node(wpo.get_exit_of_head(v)), wpo.get_node(v))
+            << wpo.get_node(v);
+        if (!first) {
+          wto << ' ';
+        }
+        wto << '(' << wpo.get_node(v);
+      } else if (wpo.is_exit(v)) {
+        EXPECT_EQ(answer.num_outer_preds, wpo.get_num_outer_preds(v).size())
+            << wpo.get_node(v);
+        EXPECT_EQ(wpo.get_node(wpo.get_head_of_exit(v)), wpo.get_node(v))
+            << wpo.get_node(v);
+        wto << ')';
+      } else {
+        if (!first) {
+          wto << ' ';
+        }
+        wto << wpo.get_node(v);
+      }
+      if (first) {
+        first = false;
+      }
+    }
+    EXPECT_EQ(wto.str(), "1 (2 (3 4) (6 7 9 8) 5) 10");
+  }
+}
+
+/*
+ * This example illustrates the effect of setting the 'lifted' in the WPO
+ * construction. While the resulting structures are both valid WPOs, only the
+ * WPO with 'lifted' set during its construction has the same WTO as
+ * Bourdoncle's when linearized. However, 'lifted' adds unnecessary orders
+ * between WPO nodes, so it must be simply used when creating a WTO from a WPO.
+ */
+TEST(WeakPartialOrderingTest, exampleFromWpoPaperIrreducible) {
+  {
+    SimpleGraph2 g;
+    g.add_edge("1", "2");
+    g.add_edge("2", "3");
+    g.add_edge("3", "2");
+    g.add_edge("3", "4");
+    g.add_edge("4", "3");
+    g.add_edge("2", "5");
+    g.add_edge("5", "4");
+    g.add_edge("1", "6");
+    g.add_edge("6", "4");
+
+    WeakPartialOrdering<std::string> wpo(
+        "1", [&g](const std::string& n) { return g.successors(n); }, false);
+
+    EXPECT_FALSE(wpo.is_from_outside("3", "4"));
+    EXPECT_FALSE(wpo.is_from_outside("2", "3"));
+    EXPECT_TRUE(wpo.is_from_outside("2", "6"));
+    EXPECT_TRUE(wpo.is_from_outside("3", "6"));
+    EXPECT_TRUE(wpo.is_from_outside("3", "5"));
+    EXPECT_FALSE(wpo.is_from_outside("2", "5"));
+
+    EXPECT_EQ(8, wpo.size());
+
+    // node, plain, head, exit, num_succs, num_preds, num_outer_preds
+    Answer lst[] = {
+        {"1", true, false, false, 2, 0, 0, false},
+        {"2", false, true, false, 2, 1, 0, false},
+        {"3", false, true, false, 1, 1, 0, true},
+        {"5", true, false, false, 1, 1, 0, false},
+        {"6", true, false, false, 1, 1, 0, false},
+        {"4", true, false, false, 1, 3, 0, false},
+        {"3", false, false, true, 1, 1, 2, false},
+        {"2", false, false, true, 0, 1, 2, false},
+    };
+
+    std::unordered_map<WpoIdx, uint32_t> count(wpo.size());
+    std::stack<WpoIdx> wl;
+    wl.push(wpo.get_entry());
+    int i = 0;
+    std::ostringstream wto;
+    bool first = true;
+    while (!wl.empty()) {
```

Comment:
This check-properties-and-print-wto-string code seems to be repeated across many tests... could it be factored out?